src/data/polynomial/laurent.lean

```diff
@@ -0,0 +1,186 @@
+begin
+  refine alg_hom.comp _ (to_finsupp_iso_alg R).to_alg_hom,
+  exact (map_domain_alg_hom R R (nat.cast_add_monoid_hom ℤ)),
+end
+
+namespace laurent_polynomial
+
+section semiring
+variables [semiring R]
+
+lemma single_zero_one_eq_one : (single 0 1 : R[T;T⁻¹]) = (1 : R[T;T⁻¹]) := rfl
+
+/-!  ### The functions `C` and `T`. -/
+/--  The ring homomorphism `C`, including `R` into the ring of Laurent polynomials over `R` as
+the constant Laurent polynomials. -/
+def C : R →+* R[T;T⁻¹] :=
+add_monoid_algebra.single_zero_ring_hom
+
+lemma single_eq_C (r : R) : single 0 r = C r := rfl
+
+/--  The function `n ↦ T ^ n`, implemented as a sequence `ℤ → R[T;T⁻¹]`.
+
+Using directly `T ^ n` does not work, since we want the exponents to be of Type `ℤ` and there
+is no `ℤ`-power defined on `R[T;T⁻¹]`.  Using that `T` is a unit introduces extra coercions.
+For these reasons, the definition of `T` is as a sequence. -/
+def T (n : ℤ) : R[T;T⁻¹] := single n 1
+
+@[simp]
+lemma T_zero : (T 0 : R[T;T⁻¹]) = 1 := rfl
+
+lemma T_add (m n : ℤ) : (T (m + n) : R[T;T⁻¹]) = T m * T n :=
+by { convert single_mul_single.symm, simp [T] }
+
+@[simp]
+lemma T_pow (m : ℤ) (n : ℕ) : (T m ^ n : R[T;T⁻¹]) = T (n * m) :=
+by rw [T, T, single_pow n, one_pow, nsmul_eq_mul, int.nat_cast_eq_coe_nat]
+
+/-- The `simp` version of `mul_assoc`, in the presence of `T`'s. -/
+@[simp]
+lemma mul_T_assoc (f : R[T;T⁻¹]) (m n : ℤ) : f * T m * T n = f * T (m + n) :=
+by simp [← T_add, mul_assoc]
+
+@[simp]
+lemma single_eq_C_mul_T (r : R) (n : ℤ) :
+  (single n r : R[T;T⁻¹]) = (C r * T n : R[T;T⁻¹]) :=
+by convert single_mul_single.symm; simp
+
+-- This lemma locks in the right changes and is what Lean proved directly.
+-- The actual `simp`-normal form of a Laurent monomial is `C a * T n`, whenever it can be reached.
+@[simp]
+lemma _root_.polynomial.to_laurent_C_mul_T (n : ℕ) (r : R) :
+  ((polynomial.monomial n r).to_laurent : R[T;T⁻¹]) = C r * T n :=
+begin
+  show map_domain (nat.cast_add_monoid_hom ℤ) ((to_finsupp_iso R) (monomial n r)) =
+    (C r * T n : R[T;T⁻¹]),
+  convert (map_domain_single : _ = single (nat.cast_add_monoid_hom ℤ n) r),
+  { exact to_finsupp_monomial n r },
+  { simp only [nat.coe_cast_add_monoid_hom, int.nat_cast_eq_coe_nat, single_eq_C_mul_T] },
+end
+
+@[simp]
+lemma _root_.polynomial.to_laurent_C (r : R) :
+  (polynomial.C r).to_laurent = C r :=
+begin
+  convert polynomial.to_laurent_C_mul_T 0 r,
+  simp only [int.coe_nat_zero, T_zero, mul_one],
+end
+
+@[simp]
+lemma _root_.polynomial.to_laurent_X :
+  (polynomial.X.to_laurent : R[T;T⁻¹]) = T 1 :=
+begin
+  have : (polynomial.X : R[X]) = monomial 1 1,
+  { simp [monomial_eq_C_mul_X] },
+  simp [this, polynomial.to_laurent_C_mul_T],
+end
+
+@[simp] lemma _root_.polynomial.to_laurent_one : (polynomial.to_laurent : R[X] → R[T;T⁻¹]) 1 = 1 :=
+map_one polynomial.to_laurent
+
+@[simp]
+lemma _root_.polynomial.to_laurent_C_mul_eq (r : R) (f : R[X]):
+  (polynomial.C r * f).to_laurent = C r * f.to_laurent :=
```

Comment:
Surely the linter rejects this, as both rewrites are themselves simp lemmas?

Reply:
The local linter on my computer did not complain.  It seems that the Eric-based linter does though: should I remove the `simp` attribute, or should I wait for CI to tell me what to do (and do nothing if there are no complaints)?

Reply:
The eric-based linter doesn't really care, but worries that the real linter doesn't do what it claims to do any more.

Reply:
Ok, I will not remove the `simp` attribute and we will see if the linter complains.

Reply:
There you go: linting passed and `simp` lemma proved by `simp`!

Reply:
Btw, also `_root_.polynomial.to_laurent_C_mul_X_pow` (the lemma following the next one) has a similar behaviour: it is a `simp` lemma that can be proved with `by simp`.